---
title: "Imputation"
author: "LucianoRogerio"
date: "2022-08-01"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Introduction

```{bash, eval = F}
cp -r /home/jj332_cas/CassavaGenotypeData/CassavaGeneticMap /workdir/lbraatz/IITA_2022GS/data/.;

```

## Impute with West Africa RefPanel

Impute with Beagle V5.0.

Use the “imputation reference panel” dataset from 2021 merged with the imputed GS progeny, e.g. chr1_RefPanelAndGSprogeny_ReadyForGP_2021Aug08.vcf.gz as reference for the current imputation.
Downloaded from [cassavabase FTP server](https://cassavabase.org/ftp/marnin_datasets/IITA_2021GS/output/)

Used cbsumezey03 Cornell CBSU machine (e.g. cbsumezey03; 20 cores, 512 GB RAM), running 1 chromosome at a time.

```{r, eval = F}
targetVCFpath<-here::here("data/Report-DCas22-7004/") # location of the targetVCF
refVCFpath<-here::here("data/")
mapPath<-here::here("data/CassavaGeneticMap/")
outPath<-here::here("output/")
outSuffix<-"DCas22_7004"

library(tidyverse); library(magrittr); 
library(genomicMateSelectR)
purrr::map(1:18,
           ~genomicMateSelectR::runBeagle5(targetVCF=paste0(targetVCFpath,"chr",.,
                                                            "_DCas22_7004.vcf.gz"),
                                           refVCF=paste0(refVCFpath,"chr",.,
                                                         "_RefPanelAndGSprogeny_ReadyForGP_2021Aug08.vcf.gz"),
                                           mapFile=paste0(mapPath,"chr",.,
                                                          "_cassava_cM_pred.v6_91019.map"),
                                           outName=paste0(outPath,"chr",.,
                                                          "_DCas22_7004_WA_REFimputed"),
                                           nthreads=20))

purrr::map(3,
           ~genomicMateSelectR::runBeagle5(window = 60, targetVCF=paste0(targetVCFpath,"chr",.,
                                                            "_DCas22_7004.vcf.gz"),
                                           refVCF=paste0(refVCFpath,"chr",.,
                                                         "_RefPanelAndGSprogeny_ReadyForGP_2021Aug08.vcf.gz"),
                                           mapFile=paste0(mapPath,"chr",.,
                                                          "_cassava_cM_pred.v6_91019.map"),
                                           outName=paste0(outPath,"chr",.,
                                                          "_DCas22_7004_WA_REFimputed"),
                                           nthreads=20))

```

Clean up Beagle log files after run. Move to sub-directory output/BeagleLogs/.

```{bash, eval = F}
cd /workdir/lbraatz/IITA_2022GS/output/; 
mkdir BeagleLogs;
cp *_DCas22_7004_WA_REFimputed.log BeagleLogs/
cp -r BeagleLogs /home/jj332_cas/lbraatz/IITA_2022GS/output/
cp *_DCas22_7004_WA_REFimputed* /home/jj332_cas/lbraatz/IITA_2022GS/output/
cp *_DCas22_7004_WA_REFimputed.vcf.gz /home/jj332_cas/lbraatz/IITA_2022GS/output/
```

##Post-impute filter

Standard post-imputation filter: AR2>0.75 (DR2>0.75 as of Beagle5.0), P_HWE>1e-20, MAF>0.005 [0.5%].

Loop to filter all 18 VCF files in parallel

```{r, eval = F}
inPath<-here::here("output/")
outPath<-here::here("output/")
require(furrr); plan(multicore, workers = 18)
future_map(1:18,
           ~genomicMateSelectR::postImputeFilter(inPath=inPath,
                                                 inName=paste0("chr",.,"_DCas22_7004_WA_REFimputed"),
                                                 outPath=outPath,
                                                 outName=paste0("chr",.,"_DCas22_7004_WA_REFimputedAndFiltered")))
plan(sequential)
```

Check what’s left

```{r, eval = F}
purrr::map(1:18,~system(paste0("zcat ",here::here("output/"),"chr",.,"_DCas22_7004_WA_REFimputedAndFiltered.vcf.gz | wc -l")))
# Chr01 - 7,321
# Chr02 - 3,517
# Chr03 - 3,606
# Chr04 - 3,050
# Chr05 - 3,644
# Chr06 - 3,324
# Chr07 - 1,633
# Chr08 - 3,082
# Chr09 - 3,208
# Chr10 - 2,484
# Chr11 - 2,803
# Chr12 - 2,700
# Chr13 - 2,478
# Chr14 - 4,911
# Chr15 - 3,414
# Chr16 - 2,672
# Chr17 - 2,429
# Chr18 - 2,732
```

```{bash, eval = F}
cd /workdir/lbraatz/IITA_2022GS/output/;
cp -r *_DCas22_7004_WA_REFimputed* /home/jj332_cas/lbraatz/IITA_2022GS/output/
```

## Formats for downstream analysis

Need to create a genome-wide VCF with the RefPanel + DCas22_7004 VCFs merged.


```{r Prepare Files to Downstream Analysis, eval=F}
require(furrr); plan(multicore, workers = 18)
# 1. Subset RefPanel to sites remaining after post-impute filter of DCas22_7004
future_map(1:18,~system(paste0("vcftools --gzvcf ",
                               "/workdir/lbraatz/IITA_2022GS/data/chr",
                               .,"_RefPanelAndGSprogeny_ReadyForGP_2021Aug08.vcf.gz"," ",
                               "--positions ","/workdir/lbraatz/IITA_2022GS/output/chr",.,
                               "_DCas22_7004_WA_REFimputed.sitesPassing"," ",
                               "--recode --stdout | bgzip -c -@ 24 > ",
                               "/workdir/lbraatz/IITA_2022GS/output/chr",.,
                               "_RefPanelAndGSprogeny2021Aug08_SubsetAndReadyToMerge.vcf.gz")))
plan(sequential)

# 2. Merge RefPanel and DCas22_7004
library(tidyverse); library(magrittr); library(genomicMateSelectR)
inPath<-here::here("output/")
outPath<-here::here("output/")
future_map(1:18,~mergeVCFs(inPath=inPath,
                           inVCF1=paste0("chr",.,"_RefPanelAndGSprogeny2021Aug08_SubsetAndReadyToMerge"),
                           inVCF2=paste0("chr",.,"_DCas22_7004_WA_REFimputedAndFiltered"),
                           outPath=outPath,
                           outName=paste0("chr",.,"_RefPanelAndGSprogeny_ReadyForGP_2022Aug03")))

# 3. Concatenate chromosomes

## Index with tabix first
future_map(1:18,~system(paste0("tabix -f -p vcf ",inPath,
                               "chr",.,"_RefPanelAndGSprogeny_ReadyForGP_2022Aug03.vcf.gz")))
plan(sequential)
## bcftools concat
system(paste0("bcftools concat ",
              "--output ",outPath,
              "AllChrom_RefPanelAndGSprogeny_ReadyForGP_2022Aug03.vcf.gz ",
              "--output-type z --threads 18 ",
              paste0(inPath,"chr",1:18,
                     "_RefPanelAndGSprogeny_ReadyForGP_2022Aug03.vcf.gz",
                     collapse = " ")))

## Remove the Old DS format of the VCF file

system(paste0("bcftools annotate -x ^INFO/PASS,^FORMAT/GT ",
              "AllChrom_RefPanelAndGSprogeny_ReadyForGP_2022Aug03.vcf.gz | ",
              "bgzip -c -@20 > ",
              "AllChrom_RefPanelAndGSprogeny_ReadyForGP_2022Aug31.vcf.gz"))

## Convert to binary blink (bed/bim/fam)
inPath<-here::here("output/")
outPath<-here::here("output/")
vcfName<-"AllChrom_RefPanelAndGSprogeny_ReadyForGP_2022Aug31"
system(paste0("export PATH=/programs/plink-1.9-x86_64-beta3.30:$PATH;",
              "plink --vcf ",inPath,vcfName,".vcf.gz ",
              "--make-bed --const-fid --keep-allele-order ",
              "--out ",outPath,vcfName))
```

